In [170]:
import pandas as pd
import altair as alt
import numpy as np
import time
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support.expected_conditions import visibility_of_element_located

scriptures = pd.read_csv("C:/Users/LeRoy12358/Documents/School/W23/499 DS - Senior Project/WWPapers/raw data/lds-scriptures.csv")

# scriptures

In [165]:
browser = webdriver.Chrome()  # start a web browser
browser.get("https://scriptures.byu.edu/#:t3d092:p573")  # navigate to URL
# wait for page to load
# by waiting for <h1> element to appear on the page
title = (
    WebDriverWait(driver=browser, timeout=10)
    .until(visibility_of_element_located((By.ID, "jContent")))
    .text
)
# retrieve fully rendered HTML content
content = browser.page_source
browser.close()



In [5]:
# # we then could parse it with beautifulsoup
soup = BeautifulSoup(content, "html.parser")
# print(soup.find("jContent").text)
# soup

soup.div['id']

# print(soup.prettify())
# content
# soup.div.div.div

soup.find("div", {"id": "jContent"})
tmp = soup.find('div', {'class': 'discourseBody'}).text

# tmp.replace("\xa0", "")
# tmp

In [66]:
links = soup.find('ul', {'class': 'talksblock'})

# 250002
# 240028
# 240007
# 230038
# 230014
# 230009
# 220046

linkList = links.find_all('a', {'class': 'refcounter'})
# .extract()
# links
linkList


link1 = links.find('a', {'class': 'refcounter'})['onclick']

# type(link1)
link1.replace("getTalk('", "").replace("');", "")
for item in linkList:
    # item.replace("getTalk('", "").replace("');", "")
    # print(item)
    pass
# linkList.replace("getTalk('", "").replace("');", "")

linkList2 = []

for link in soup.find_all('a', {'class': 'refcounter'}):
    linkList2.append(
        np.base_repr(int(
        link.get('onclick').replace("getTalk('", "").replace("');", "")
        ), 16
        ).lower()
        )
# http://example.com/elsie
# http://example.com/lacie
# http://example.com/tillie

linkList2

['3d092',
 '3a99c',
 '3a987',
 '38296',
 '3827e',
 '38279',
 '35b8e',
 '35b8d',
 '35b8b',
 '35b75',
 '35b7e',
 '35b72',
 '35b79',
 '33473',
 '33471',
 '3346f',
 '33466',
 '3345f',
 '2e662',
 '2e65a',
 '2e648',
 '2e653',
 '2bf3b',
 '2bf36',
 '2bf2f',
 '2bf2e',
 '2bf25',
 '29834',
 '2982e',
 '2981b',
 '27125',
 '27106',
 '24a1a',
 '24a13',
 '249fc',
 '249f2',
 '222e1',
 '222e6',
 '1fbe3',
 '1fbf5',
 '1d4f7',
 '1d4c4',
 '1ade7',
 '1add5',
 '1adba',
 '186cc',
 '186a3',
 '15fd2',
 '15faf',
 '15f9c',
 '138c6',
 '138c5',
 '1117f',
 'ea75',
 'ea72',
 'c37b',
 'c363',
 '9c7a',
 'c35b',
 '9c6c',
 '9c66',
 '9c5f',
 '9c54',
 '15fb9',
 '4e41']

The try block lets you test a block of code for errors.

The except block lets you handle the error.

The else block lets you execute code when there is no error.

The finally block lets you execute code, regardless of the result of the try- and except blocks.

In [179]:
# This will likely take quite a while to run....

# breaker = False
browser = webdriver.Chrome()  # start a web browser
pot = []

for i in range(0, len(linkList2)):
    url = linkList2[i]
    print(url)
    
    try:


        browser.get("https://scriptures.byu.edu/#:t" + url + ":p573")  # navigate to URL
        # wait for page to load by waiting for <h1> element to appear on the page
        title = (
            WebDriverWait(driver=browser, timeout=10)
            .until(visibility_of_element_located((By.ID, "jContent")))
            .text
        )
        # retrieve fully rendered HTML content
        content = browser.page_source

    except:
        print("Something broke but I might have fixed it :D")
        browser.close()
        browser = webdriver.Chrome()
        i -= 1
        # continue
    else:
        # we then could parse it with beautifulsoup
        soup = BeautifulSoup(content, "html.parser")
        pot.append(soup)
        time.sleep(.5)
    
browser.close()
    # if(breaker):
    #     break
    # breaker = True
    

3d092
3a99c
Something broke but I might have fixed it :D
3a987
38296
Something broke but I might have fixed it :D
3827e
38279
Something broke but I might have fixed it :D
35b8e
Something broke but I might have fixed it :D


WebDriverException: Message: disconnected: not connected to DevTools
  (failed to check if window was closed: disconnected: not connected to DevTools)
  (Session info: chrome=110.0.5481.178)
Stacktrace:
Backtrace:
	(No symbol) [0x00FB37D3]
	(No symbol) [0x00F48B81]
	(No symbol) [0x00E4B36D]
	(No symbol) [0x00E3BE5A]
	(No symbol) [0x00E3C63A]
	(No symbol) [0x00E3C5D8]
	(No symbol) [0x00E2D0BA]
	(No symbol) [0x00E2DBA7]
	(No symbol) [0x00EA489B]
	(No symbol) [0x00E9B3E3]
	(No symbol) [0x00E70D97]
	(No symbol) [0x00E7253D]
	GetHandleVerifier [0x0122ABF2+2510930]
	GetHandleVerifier [0x01258EC1+2700065]
	GetHandleVerifier [0x0125C86C+2714828]
	GetHandleVerifier [0x01063480+645344]
	(No symbol) [0x00F50FD2]
	(No symbol) [0x00F56C68]
	(No symbol) [0x00F56D4B]
	(No symbol) [0x00F60D6B]
	BaseThreadInitThunk [0x778300F9+25]
	RtlGetAppContainerNamedObjectPath [0x77EA7BBE+286]
	RtlGetAppContainerNamedObjectPath [0x77EA7B8E+238]


['3d092', '3a99c', '3a987', '38296', '3827e', '38279', '35b8e', '35b8d', '35b8b', '35b75', '35b7e', '35b72', '35b79', '33473', '33471', '3346f', '33466', '3345f',
Output exceeds the size limit. Open the full output data in a text editor
3d092, 3a99c, Something broke but I might have fixed it :D, 3a987, 38296, Something broke but I might have fixed it :D, 3827e, 38279, 35b8e, Something broke but I might have fixed it :D, 35b8d, 35b8b, Something broke but I might have fixed it :D, 35b75, 35b7e, 35b72, 35b79, 33473

In [94]:
# pot.pop()
pot2 = pot
# pot.pop()
pot2.__sizeof__()
pot.__sizeof__()

40

In [160]:
from difflib import SequenceMatcher

def findMatches(str1, str2, BLOCK_MIN_LENGTH = 6):
    sm = SequenceMatcher(lambda x: x in " ", str1, str2)
    finalMatches = []
    matchesPosns = sm.get_matching_blocks()

    # put together a list of matches longer than only a couple characters
    for matchPos in matchesPosns:
        matchLength = matchPos[2]
        if matchLength < BLOCK_MIN_LENGTH:
            continue
        # print(str1[matchPos[0]:matchPos[0] + matchPos[2]])
        finalMatches.append(matchPos)
    return [str1[finalMatches[0][0]:finalMatches[-1][0] + finalMatches[-1][2]],
            str2[finalMatches[0][1]:finalMatches[-1][1] + finalMatches[-1][2]]]


In [161]:
myStr1 = "He has left on record through the medium of holy men who wrote and spoke as they were moved upon by the Holy Ghost"
myStr2 = "For the prophecy came not in old time by the will of man: but holy men of God spake as they were moved by the Holy Ghost."
findMatches(myStr1, myStr2)




[' holy men who wrote and spoke as they were moved upon by the Holy Ghost',
 ' holy men of God spake as they were moved by the Holy Ghost']